src/Model_Hydro/MHD_Init_BField_ByFile.cpp

```diff
@@ -0,0 +1,419 @@
+   hid_t dataset, dataspace;
+   hsize_t dims[3], maxdims[3];
+
+   hid_t mag_file_id = H5Fopen(B_Filename, H5F_ACC_RDONLY, H5P_DEFAULT);
+
+   if ( B_lv == 0 ) {
+
+     dataset = H5Dopen(mag_file_id, "magnetic_vector_potential_z", H5P_DEFAULT);
+     
+     dataspace = H5Dget_space(dataset);
+   
+     H5Sget_simple_extent_dims(dataspace, dims, maxdims);
+
+     H5Sclose(dataspace);
+     H5Dclose(dataset);
+
+//   NOTE: Magnetic vector potential arrays are stored in column-major order,
+//   i.e., Ax[nAx][nAy][nAz]
+
+     nAx = dims[0];
+     nAy = dims[1];
+     nAz = dims[2];
+
+   }
+
+#  endif
+
+// Read the coordinate information from the vector potential grid
+   Axcoord = new double [ nAx ];
+   Aycoord = new double [ nAy ];
+   Azcoord = new double [ nAz ];
+
+#  ifdef SUPPORT_HDF5
+
+   dataset = H5Dopen(mag_file_id, "x", H5P_DEFAULT);
+   status  = H5Dread(dataset, H5T_NATIVE_DOUBLE, H5S_ALL,
+      H5S_ALL, H5P_DEFAULT, Axcoord);
+   H5Dclose(dataset);
+
+   dataset = H5Dopen(mag_file_id, "y", H5P_DEFAULT);
+   status  = H5Dread(dataset, H5T_NATIVE_DOUBLE, H5S_ALL,
+      H5S_ALL, H5P_DEFAULT, Aycoord);
+   H5Dclose(dataset);
+
+   dataset = H5Dopen(mag_file_id, "z", H5P_DEFAULT);
+   status  = H5Dread(dataset, H5T_NATIVE_DOUBLE, H5S_ALL,
+      H5S_ALL, H5P_DEFAULT, Azcoord);
+   H5Dclose(dataset);
+
+#  endif
+
+// Cell spacing and left edge of vector potential grid
+
+   Adx = Axcoord[1]-Axcoord[0];
+   Ady = Aycoord[1]-Aycoord[0];
+   Adz = Azcoord[1]-Azcoord[0];
+
+   Axmin = Axcoord[0]-0.5*Adx;
+   Aymin = Aycoord[0]-0.5*Ady;
+   Azmin = Azcoord[0]-0.5*Adz;
+
+   double Axmax = Axcoord[nAx-1]+0.5*Adx;
+   double Aymax = Aycoord[nAy-1]+0.5*Ady;
+   double Azmax = Azcoord[nAz-1]+0.5*Adz;
+
+   if ( amr->BoxEdgeL[0] < Axmin || amr->BoxEdgeR[0] > Axmax ||
+        amr->BoxEdgeL[1] < Aymin || amr->BoxEdgeR[1] > Aymax ||
+        amr->BoxEdgeL[2] < Azmin || amr->BoxEdgeR[2] > Azmax ) 
+      Aux_Error( ERROR_INFO, "Input grid is smaller than the simulation domain !!\n" ); 
+
+   double *Ax = new double [ CUBE(PS1+1) ];
+   double *Ay = new double [ CUBE(PS1+1) ];
+   double *Az = new double [ CUBE(PS1+1) ];
+
+   double sample_res = pow(2, MAX_LEVEL-B_lv);
+   double sample_fact = 1.0/((double)sample_res);
+
+   for (int PID=0; PID<amr->NPatchComma[B_lv][1]; PID++) {
+
+      double EdgeL[3];
+      double EdgeR[3];
+
+      for (int i=0; i<3; i++) { 
+         EdgeL[i] = amr->patch[0][B_lv][PID]->EdgeL[i];
+         EdgeR[i] = amr->patch[0][B_lv][PID]->EdgeR[i];
+      }
+
+//    Compute the beginning and ending indices on the vector potential grid
+//    +/- 1 are necessary because we will be computing derivatives
+
+      int ibegin = (int)((EdgeL[0]-Axmin)/Adx)-1;
+      int jbegin = (int)((EdgeL[1]-Aymin)/Ady)-1;
+      int kbegin = (int)((EdgeL[2]-Azmin)/Adz)-1;
+      
+      int iend   = (int)((EdgeR[0]-Axmin)/Adx)+1;
+      int jend   = (int)((EdgeR[1]-Aymin)/Ady)+1;
+      int kend   = (int)((EdgeR[2]-Azmin)/Adz)+1;
```

Comment:
Load two instead of one buffer cells. It is to avoid the **index out of range** problem resulting from round-off errors when the vector potential grid and AMR blocks align almost perfectly with each other.
```
      int ibegin = (int)((EdgeL[0]-Axmin)/Adx)-2;
      int jbegin = (int)((EdgeL[1]-Aymin)/Ady)-2;
      int kbegin = (int)((EdgeL[2]-Azmin)/Adz)-2;

      int iend   = (int)((EdgeR[0]-Axmin)/Adx)+2;
      int jend   = (int)((EdgeR[1]-Aymin)/Ady)+2;
      int kend   = (int)((EdgeR[2]-Azmin)/Adz)+2;
```